ykllvmwrap/src/ykllvmwrap.cc

```diff
@@ -146,10 +146,27 @@ extern "C" void *__ykllvmwrap_irtrace_compile(char *FuncNames[], size_t BBs[],
           Tracing = false;
         }
       }
-      if (!Tracing && !llvm::isa<AllocaInst>(I)) {
-        // FIXME Remove allocas not needed in traced code.
+
+      if (!Tracing)
         continue;
+
+      // FIXME For any variable defined outside of the trace, we realise
+      // dummy storage space. Eventually, such variables should be inputs
+      // to the trace. This hack prevents the remapper from choking when
+      // it can't find the definition of an instruction operand.
+      if ((llvm::isa<llvm::StoreInst>(I)) && (VMap[&*I] == nullptr)) {
+        Value *DestOp = I->getOperand(1);
+        Value *SrcOp = I->getOperand(0);
+        Instruction *Alloca =
+            Builder.CreateAlloca(SrcOp->getType(), (unsigned)0);
```

Comment:
Does it make a difference that these allocas are spread through-out the block[s]. Not sure if it's worth it moving them all into the beginning of first one, similar to what LLVM does.

Reply:
The module verified, so I think it's ok. Since this isn't going to be the final design, I don't think it's worth worrying about.

Reply:
Alloca's outside of the first block will dynamically grow the stackframe each time they are encountered and grow the stackframe multiple times if they are hit multiple times. Alloca's inside the first block will all be grouped together to grow the stackframe in the prologue once.

Reply:
I'd expect optimisation passes to push allocations up?

Reply:
Nope, even with `-O3` I get:

```llvm
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
target triple = "x86_64-unknown-linux-gnu"

define void @first_alloca() unnamed_addr #0 {
  %baz = alloca i8, align 1
  br label %bb0

bb0:                       
  store i8 0, i8* %baz, align 1
  call void @foo(i8* noalias readonly align 1 dereferenceable(1) %baz)
  br label %bb1

bb1:                                              ; preds = %bb0
  br label %bb0
}

define void @second_alloca() unnamed_addr #0 {
  br label %bb0

bb0:                       
  %baz = alloca i8, align 1
  store i8 0, i8* %baz, align 1
  call void @foo(i8* noalias readonly align 1 dereferenceable(1) %baz)
  br label %bb1

bb1:                                              ; preds = %bb0
  br label %bb0
}

declare void @foo(i8* noalias readonly align 1 dereferenceable(1)) unnamed_addr

attributes #0 = { nonlazybind uwtable "probe-stack"="__rust_probestack" "target-cpu"="x86-64" }
```

```asm
first_alloca:                           # @first_alloca
        push    rbx
        sub     rsp, 16
        lea     rbx, [rsp + 15]
.LBB0_1:                                # %bb0
        mov     byte ptr [rsp + 15], 0
        mov     rdi, rbx
        call    foo@PLT
        jmp     .LBB0_1
second_alloca:                          # @second_alloca
        push    rbp
        mov     rbp, rsp
.LBB1_1:                                # %bb0
        mov     eax, 16
        call    __rust_probestack
        sub     rsp, rax
        mov     rdi, rsp
        mov     byte ptr [rdi], 0
        call    foo@PLT
        jmp     .LBB1_1
```

Reply:
Huh! Surprising!

Reply:
LLVM is optimized for the code patterns clang uses. Clang places alloca's at the first block for regular variables, so I am not too surprised that LLVM doesn't optimize alloca's in other blocks.